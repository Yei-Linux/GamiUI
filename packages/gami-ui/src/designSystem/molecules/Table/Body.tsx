import { IDynamicObjectWithField } from 'core/domain/interfaces/common'
import React from 'react'
import * as S from './Table.styles'

export type IBodyChildren = (
  item: IDynamicObjectWithField,
  index: number
) => React.ReactNode

export interface IBody {
  children: IBodyChildren
  items: IDynamicObjectWithField[]
}

export const Body = ({ children, items }: IBody) => {
  return <S.TableBody>{items.map(children)}</S.TableBody>
}
