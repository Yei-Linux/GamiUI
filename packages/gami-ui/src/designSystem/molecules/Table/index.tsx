export { default } from './Table'
export { options } from './constants'
