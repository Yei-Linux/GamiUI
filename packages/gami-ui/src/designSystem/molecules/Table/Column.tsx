import { cls } from 'core/utils/cls'
import React from 'react'
import * as S from './Table.styles'

export interface IColumn {
  children: React.ReactNode
  as?: 'th' | 'td'
  colspan?: number
}

export const Column = ({ children, as = 'td', colspan }: IColumn) => {
  return (
    <S.TableColumn
      as={as}
      colSpan={colspan}
      className={cls({
        header: as === 'th',
      })}
    >
      {children}
    </S.TableColumn>
  )
}
