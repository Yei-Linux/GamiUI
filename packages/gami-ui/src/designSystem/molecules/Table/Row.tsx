import { IDynamicObjectWithField } from 'core/domain/interfaces/common'
import React from 'react'
import { defaultTheme } from 'styles/tokens'
import * as S from './Table.styles'

export type IRowChildren = (
  entry: [key: string, value: unknown],
  index: number
) => React.ReactNode

export interface IRow {
  item?: IDynamicObjectWithField
  children: React.ReactNode | IRowChildren
  background?: string
}

export const Row = ({
  children,
  item,
  background = defaultTheme.light.neutral[700],
}: IRow) => {
  if (!item)
    return <S.TableRow $background={background}>{children as any}</S.TableRow>

  if (typeof children !== 'function') return null

  return (
    <S.TableRow $background={background}>
      {Object.entries(item).map(children as IRowChildren)}
    </S.TableRow>
  )
}
