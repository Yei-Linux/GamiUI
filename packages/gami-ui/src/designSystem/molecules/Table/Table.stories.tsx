import React from 'react'
import { Meta } from '@storybook/react'

import Table from '.'
import RichText from '../../atoms/RichText'

export default {
  title: 'Molecules/Table',
  component: Table,
  args: {},
  argTypes: {},
} as Meta

export const Basic = () => {
  const columns = [
    { title: 'NOMBRES', dataIndex: 'name' },
    { title: 'EDAD', dataIndex: 'age' },
    { title: 'DIRECCION', dataIndex: 'address' },
    { title: 'PAIS', dataIndex: 'country' },
  ]
  const items = [
    { name: 'Pedro', age: '12', address: 'Jr Junior', country: 'USA' },
    { name: 'Pedro', age: '12', address: 'Jr Junior', country: 'USA' },
  ]

  return (
    <Table>
      <Table.Header columns={columns}>
        {(column) => <Table.Column as="th">{column.title}</Table.Column>}
      </Table.Header>
      <Table.Body items={items}>
        {(item, index) => (
          <Table.Row item={item} key={index}>
            {([cellKey, cellValue], index) => (
              <Table.Cell
                key={index}
                cellKey={cellKey}
                cellValue={`${cellValue}`}
                columns={columns}
              />
            )}
          </Table.Row>
        )}
      </Table.Body>
    </Table>
  )
}

export const WithRender = () => {
  const columns = [
    {
      title: 'Nombres',
      dataIndex: 'name',
      render: (name: string) => (
        <RichText text={`[${name}](https://google.com?target=_blank)`} />
      ),
    },
    { title: 'Edad', dataIndex: 'age' },
    { title: 'Direccion', dataIndex: 'address' },
    { title: 'Pais', dataIndex: 'country' },
  ]
  const items = [
    { name: 'Pedro', age: '12', address: 'Jr Junior', country: 'USA' },
    { name: 'Pedro', age: '12', address: 'Jr Junior', country: 'USA' },
  ]

  return (
    <Table>
      <Table.Header columns={columns}>
        {(column) => <Table.Column as="th">{column.title}</Table.Column>}
      </Table.Header>
      <Table.Body items={items}>
        {(item, index) => (
          <Table.Row item={item} key={index}>
            {([cellKey, cellValue], index) => (
              <Table.Cell
                key={index}
                cellKey={cellKey}
                cellValue={`${cellValue}`}
                columns={columns}
              />
            )}
          </Table.Row>
        )}
      </Table.Body>
    </Table>
  )
}

export const WithPagination = () => {
  const columns = [
    { title: 'NOMBRES', dataIndex: 'name' },
    { title: 'EDAD', dataIndex: 'age' },
    { title: 'DIRECCION', dataIndex: 'address' },
    { title: 'PAIS', dataIndex: 'country' },
  ]
  const items = [
    { name: 'Pedro', age: '12', address: 'Jr Junior', country: 'USA' },
    { name: 'Pedro', age: '12', address: 'Jr Junior', country: 'USA' },
  ]

  const handleChangePage = (page: number) => {
    console.log('test', page)
  }

  return (
    <Table>
      <Table.Header columns={columns} background={'#fbfbfb'}>
        {(column) => <Table.Column as="th">{column.title}</Table.Column>}
      </Table.Header>
      <Table.Body items={items}>
        {(item, index) => (
          <Table.Row item={item} key={index} background={'#fbfbfb'}>
            {([cellKey, cellValue], index) => (
              <Table.Cell
                key={index}
                cellKey={cellKey}
                cellValue={`${cellValue}`}
                columns={columns}
              />
            )}
          </Table.Row>
        )}
      </Table.Body>
      <Table.Footer>
        <Table.Pagination
          columns={columns.length}
          numberPages={6}
          initialPage={0}
          onChangePage={handleChangePage}
        />
      </Table.Footer>
    </Table>
  )
}
