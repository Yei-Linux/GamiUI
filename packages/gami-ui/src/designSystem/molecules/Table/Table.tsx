import React from 'react'
import { IGeneralProps } from 'core/domain/interfaces/IGeneralProps'
import { IDynamicObjectWithField } from 'core/domain/interfaces/common'
import { getGenericPropStyles } from 'styles/utilities/genericPropStyles'

import * as S from './Table.styles'
import RichText from 'designSystem/atoms/RichText'
import { cls } from 'core/utils/cls'

interface IColumn {
  title: string
  dataIndex: string
  render?: (props: string) => React.ReactNode
}

export interface ITable extends IGeneralProps {
  columns: IColumn[]
  data: IDynamicObjectWithField[]
}

const Table = ({ columns, data, ...genericsProps }: ITable) => {
  const findRenderByColumn = (key: string, value: string) => {
    const column: IColumn | undefined = columns.find(
      ({ dataIndex }) => dataIndex == key
    )
    return column?.render ? column?.render(value) : <RichText text={value} />
  }

  return (
    <S.Table
      className={cls(genericsProps?.className ?? '')}
      {...getGenericPropStyles(genericsProps)}
    >
      <S.TableContainer>
        <S.TableHeader>
          <S.TableHeaderRow>
            {columns.map((column: IColumn, index: number) => (
              <S.TableHeaderColumn key={index}>
                {column.title}
              </S.TableHeaderColumn>
            ))}
          </S.TableHeaderRow>
        </S.TableHeader>

        <S.TableBody>
          {data.map((row: any, index: number) => (
            <S.TableBodyRow key={index}>
              {Object.keys(row).map((column: any, index: number) => (
                <S.TableBodyColumn key={index}>
                  {findRenderByColumn(column, row[column])}
                </S.TableBodyColumn>
              ))}
            </S.TableBodyRow>
          ))}
        </S.TableBody>
      </S.TableContainer>
    </S.Table>
  )
}

Table.displayName = 'Table'

export default Table
