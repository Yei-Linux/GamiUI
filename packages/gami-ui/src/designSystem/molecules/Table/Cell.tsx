import RichText from 'designSystem/atoms/RichText'
import React from 'react'
import { Column } from './Column'
import { IColumn } from './Table'

export interface ICell {
  cellKey: string
  cellValue: string
  columns: IColumn[]
}

export const Cell = ({ cellKey, cellValue, columns }: ICell) => {
  const findRenderByColumn = (
    cellKey: string,
    cellValue: string,
    columns: IColumn[]
  ) => {
    const column: IColumn | undefined = columns.find(
      ({ dataIndex }) => dataIndex == cellKey
    )
    return column?.render ? (
      column?.render(cellValue)
    ) : (
      <RichText text={cellValue} />
    )
  }

  return <Column>{findRenderByColumn(cellKey, cellValue, columns)}</Column>
}
