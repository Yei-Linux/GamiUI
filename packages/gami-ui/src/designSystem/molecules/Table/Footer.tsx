import React from 'react'
import { Row } from './Row'
import * as S from './Table.styles'

export interface IFooter {
  children: React.ReactNode
}

export const Footer = ({ children }: IFooter) => {
  return (
    <S.TableFooter>
      <Row background="transparent">{children}</Row>
    </S.TableFooter>
  )
}
