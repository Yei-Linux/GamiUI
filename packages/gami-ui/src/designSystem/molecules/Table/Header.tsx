import React from 'react'
import { defaultTheme } from 'styles/tokens'
import { Row } from './Row'
import { IColumn } from './Table'
import * as S from './Table.styles'

export type IHeaderChildren = (
  column: IColumn,
  index: number
) => React.ReactNode

export interface IHeader {
  columns: IColumn[]
  children: IHeaderChildren
  background?: string
}

export const Header = ({
  children,
  columns,
  background = defaultTheme.light.neutral[700],
}: IHeader) => {
  return (
    <S.TableHeader $background={background}>
      <Row>{columns.map(children)}</Row>
    </S.TableHeader>
  )
}
