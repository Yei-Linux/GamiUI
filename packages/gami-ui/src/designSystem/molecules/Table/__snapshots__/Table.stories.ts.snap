// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Molecules/Table Basic 1`] = `
<div
  className="css-197ljk5-TableWrapper ejhrnoo0"
>
  <table
    className="css-1v7co3z-TableContainer ejhrnoo1"
  >
    <thead
      className="css-1aolgnl-TableHeader ejhrnoo2"
    >
      <tr
        className="css-1o444mx-TableHeaderRow ejhrnoo3"
      >
        <th
          className="css-1aw9a9f-TableHeaderColumn ejhrnoo4"
        >
          Nombres
        </th>
        <th
          className="css-1aw9a9f-TableHeaderColumn ejhrnoo4"
        >
          Edad
        </th>
        <th
          className="css-1aw9a9f-TableHeaderColumn ejhrnoo4"
        >
          Direccion
        </th>
        <th
          className="css-1aw9a9f-TableHeaderColumn ejhrnoo4"
        >
          Pais
        </th>
      </tr>
    </thead>
    <tbody
      className="css-r22c21-TableBody ejhrnoo5"
    >
      <tr
        className="css-1xm5vfz-TableBodyRow ejhrnoo6"
      >
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          Pedro
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          12
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          Jr Junior
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          USA
        </td>
      </tr>
      <tr
        className="css-1xm5vfz-TableBodyRow ejhrnoo6"
      >
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          Pedro
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          12
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          Jr Junior
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          USA
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Storyshots Molecules/Table With Render 1`] = `
<div
  className="css-197ljk5-TableWrapper ejhrnoo0"
>
  <table
    className="css-1v7co3z-TableContainer ejhrnoo1"
  >
    <thead
      className="css-1aolgnl-TableHeader ejhrnoo2"
    >
      <tr
        className="css-1o444mx-TableHeaderRow ejhrnoo3"
      >
        <th
          className="css-1aw9a9f-TableHeaderColumn ejhrnoo4"
        >
          Nombres
        </th>
        <th
          className="css-1aw9a9f-TableHeaderColumn ejhrnoo4"
        >
          Edad
        </th>
        <th
          className="css-1aw9a9f-TableHeaderColumn ejhrnoo4"
        >
          Direccion
        </th>
        <th
          className="css-1aw9a9f-TableHeaderColumn ejhrnoo4"
        >
          Pais
        </th>
      </tr>
    </thead>
    <tbody
      className="css-r22c21-TableBody ejhrnoo5"
    >
      <tr
        className="css-1xm5vfz-TableBodyRow ejhrnoo6"
      >
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          <div
            className="css-cbj4y6-RichTextWrapper e17e5g740"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<p><a style=\\"font-weight: bold;color: black;\\" href=\\"https://google.com\\" target=_blank>Pedro</a></p>
",
              }
            }
          />
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          12
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          Jr Junior
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          USA
        </td>
      </tr>
      <tr
        className="css-1xm5vfz-TableBodyRow ejhrnoo6"
      >
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          <div
            className="css-cbj4y6-RichTextWrapper e17e5g740"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<p><a style=\\"font-weight: bold;color: black;\\" href=\\"https://google.com\\" target=_blank>Pedro</a></p>
",
              }
            }
          />
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          12
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          Jr Junior
        </td>
        <td
          className="css-29fhvk-TableBodyColumn ejhrnoo7"
        >
          USA
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
